### PR TITLE
implement map function for list-encoding test

### PR DESCRIPTION
**Pull request by olhotak**
This is an attempt to implement map for the list in the `list-encoding` test. Currently this crashes the compiler:
```
Fatal compiler crash when compiling: tests/pos-special/captures/list-encoding.scala:
TermParamRef(op) (of class dotty.tools.dotc.core.Types$TermParamRefImpl)
dotty.tools.dotc.typer.CheckCaptures.traverse$$anonfun$3$$anonfun$2(CheckCaptures.scala:126)
scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
dotty.tools.dotc.util.SimpleIdentitySet$Set1.foreach(SimpleIdentitySet.scala:69)
dotty.tools.dotc.typer.CheckCaptures.dotty$tools$dotc$typer$CheckCaptures$PostRefinerCheck$$$_$traverse$$anonfun$1(CheckCaptures.scala:136)
scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
scala.collection.immutable.List.foreach(List.scala:333)
dotty.tools.dotc.typer.CheckCaptures$PostRefinerCheck$.traverse(CheckCaptures.scala:136)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1631)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1631)
dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1525)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1632)
dotty.tools.dotc.typer.CheckCaptures$PostRefinerCheck$.traverse(CheckCaptures.scala:147)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1631)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1631)
dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1519)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1632)
dotty.tools.dotc.typer.CheckCaptures$PostRefinerCheck$.traverse(CheckCaptures.scala:147)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1631)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1631)
dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1525)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1632)
dotty.tools.dotc.typer.CheckCaptures$PostRefinerCheck$.traverse(CheckCaptures.scala:147)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1631)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1631)
dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1594)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1632)
dotty.tools.dotc.typer.CheckCaptures$PostRefinerCheck$.traverse(CheckCaptures.scala:147)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1631)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1631)
dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.fold$1(Trees.scala:1506)
dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.apply(Trees.scala:1508)
dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1601)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1632)
dotty.tools.dotc.typer.CheckCaptures$PostRefinerCheck$.traverse(CheckCaptures.scala:147)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1631)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1631)
dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1598)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1632)
dotty.tools.dotc.typer.CheckCaptures$PostRefinerCheck$.traverse(CheckCaptures.scala:147)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1631)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1631)
dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.fold$1(Trees.scala:1506)
dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.apply(Trees.scala:1508)
dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1607)
dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1632)
dotty.tools.dotc.typer.CheckCaptures$PostRefinerCheck$.traverse(CheckCaptures.scala:147)
dotty.tools.dotc.typer.CheckCaptures.postRefinerCheck(CheckCaptures.scala:150)
dotty.tools.dotc.typer.RefineTypes.run(RefineTypes.scala:66)
dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:308)
```

### CODE REVIEW

**Comment by olhotak** [comment]
Now it compiles after making `isGlobal` return `false` for a `TermParamRef`.

File: compiler/src/dotty/tools/dotc/typer/CheckCaptures.scala [hunk 1]
```diff
@@ -124,6 +124,7 @@ class CheckCaptures extends RefineTypes:
               val isGlobal = ref match
                 case ref: TypeRef => ref.isRootCapability
                 case ref: TermRef => ref.prefix != NoPrefix && ref.symbol.hasAnnotation(defn.AbilityAnnot)
+                case ref: TermParamRef => false
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/CheckCaptures.scala:127]: I suggest
```suggestion
                case _ => false
```
TypeParamRefs and ThisTypes are also members of capture sets

**Comment by olhotak** [comment]
After trying to add an implementation of `lazyMap`, I get the compiler assertion failure:
```
Fatal compiler crash when compiling: tests/pos-special/captures/list-encoding.scala:
assertion failed: Denotation some [C <: Top](op: Op[(Unit => A) retains A, C]): (s: C) => C retains op of selection xs.apply with type (xs.apply : [C <: Top](op: Op[(Unit => A) retains A, C]): (s: C) => C retains op) does not have a symbol
scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
dotty.tools.dotc.transform.TreeChecker$Checker.typedSelect(TreeChecker.scala:385)
dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2683)
dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2775)
dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:118)
dotty.tools.dotc.transform.TreeChecker$Checker.typedUnadapted(TreeChecker.scala:326)
```

**Comment by olhotak** [comment]
Minimized to the following:
```scala
def withRetains(f: ([T] => Int => Int) retains *) = f.apply(5)
def withoutRetains(f: ([T] => Int => Int)) = f.apply(5)
def monomorphic(f: (Int => Int) retains *) = f.apply(5)
```
`withRetains` crashes:
```
*** error while checking tests/pos-special/captures/list-encoding.scala after phase typer ***
java.lang.AssertionError: assertion failed: Denotation some [T](x$1: Int): Int of selection f.apply with type (f.apply : [T](x$1: Int): Int) does not have a symbol while compiling tests/pos-special/captures/list-encoding.scala
```

`withoutRetains` and `monomorphic` compile fine.

**Comment by olhotak** [comment]
The implementation of lazymap matching what we had during the arxiv work now compiles and runs correctly.

@odersky: the change in compiler/src/dotty/tools/dotc/transform/TreeChecker.scala works in that it removes the assertion, but is it the correct change?

The types in `lazyMap` match what we had before, but it's possible they capture more than they have to. I'm sure there are other things that could be cleaned up.

**Review by odersky — APPROVED** [review]

**Review by odersky — COMMENTED** [review]
`list-encoding.scala` needs to be moved to `pos-custom-args/captures`.

Otherwise LGTM

**Comment by olhotak** [comment]
After rebasing I get a Ycheck failure and I'm stuck on debugging it. I've minimized `list-encoding.scala` to focus on the failure.

The problem is that after phase `cc`, `Closure` trees still have their original types without `retains`. Yet when I add tracing to `CheckCaptures.typedClosure`, I see them being given types with `retains`.

The Ycheck output is:
```
*** error while checking tests/pos-custom-args/captures/list-encoding.scala after phase cc ***
java.lang.AssertionError: assertion failed: Types differ
Original type : A => List[B] => List[B]
After checking: (hd: A) => (List[B] => List[B]) retains strictMapF retains hd
Original tree : closure($anonfun)
After checking: closure($anonfun)
```

The tree (with some surrounding context) is:
```
    def strictMap[A >: Nothing <: Top, B >: Nothing <: Top](xs: List[A])(strictMapF: (A => B) retains *): List[B] = 
      xs.apply[List[B]](
        {
          def $anonfun(hd: A): List[B] => List[B] = 
            {
              def $anonfun(tl: List[B]): List[B] = cons[B](strictMapF.apply(hd), tl)
              closure($anonfun)
            }
          closure($anonfun)
        }
      ).apply(nil[B])
```

(Based on the type, I infer it's the second `closure($anonfun)`).

**Comment by olhotak** [comment]
From the trace of `CheckCaptures.typedClosure`, I see:

```
typing closure method $anonfun with fvs {strictMapF} at closure($anonfun)
capture set of (hd: A) => (List[B] => List[B]) retains strictMapF retains hd = {}
closure's type is ((hd: A) => (List[B] => List[B]) retains strictMapF retains hd) retains strictMapF
```
Notice that this type differs from both the original type and after checking type in Ycheck: compared to the after-checking type, it has the extra outer `retains strictMapF`.

**Comment by olhotak** [comment]
I traced it down to the following. The tree that is being Ychecked after the cc phase is the tree **before** it is transformed by the phase (but with symbol info from after the phase). The tree **after** being transformed by the cc phase does have the correct refined types. So it's obvious why Ycheck complains: the tree it is checking is the one without the refined types. But I don't understand how this scheme could ever work, for any program: wouldn't Ycheck always complain that it is seeing the unrefined types?

By "the tree that is being Ychecked", I mean the tree in line 147 of TreeChecker.scala:
```
    try checker.typedExpr(ctx.compilationUnit.tpdTree)(using checkingCtx)
```

**Comment by olhotak** [comment]
Turning off Ycheck for the cc phase only causes the same Ycheck failure to show up later, in phase inlinedPositions. So it looks like RefineTypes is not correctly resetting all denotations after itself. I am turning off Ycheck globally temporarily to get the list examples working.

**Comment by anatoliykmetyuk** [comment]
What's the status of this PR? Is it still being worked on? Or can we close it? @olhotak @odersky

**Comment by odersky** [comment]
I think it would be good to rebase on the current add-cc2 branch and try again.

**Comment by olhotak** [comment]
Replaced by https://github.com/dotty-staging/dotty/pull/21

**Pull request closed without merging**